src/Raven.Server/Documents/ETL/Providers/Raven/RavenEtlDocumentTransformer.cs

```diff
@@ -1,21 +1,71 @@
+using Sparrow.Json.Parsing;
+// ReSharper disable ForCanBeConvertedToForeach
 
 namespace Raven.Server.Documents.ETL.Providers.Raven
 {
     public class RavenEtlDocumentTransformer : EtlTransformer<RavenEtlItem, ICommandData>
     {
+        private readonly ScriptInput _script;
         private readonly List<ICommandData> _commands = new List<ICommandData>();
-        private readonly PatchRequest _transformationScript;
 
-        public RavenEtlDocumentTransformer(DocumentDatabase database, DocumentsOperationContext context, RavenEtlConfiguration configuration) : base(database, context)
+        private RavenEtlItem _currentlyTransformed;
+
+        public RavenEtlDocumentTransformer(DocumentDatabase database, DocumentsOperationContext context, ScriptInput script) : base(database, context)
         {
-            if (string.IsNullOrEmpty(configuration.Script) == false)
-                _transformationScript = new PatchRequest { Script = configuration.Script };
+            _script = script;
+
+            LoadToDestinations = _script.Transformation == null ? new string[0] : _script.AllCollections;
+        }
+
+        protected override string[] LoadToDestinations { get; }
+
+        protected override void LoadToFunction(string collectionName, JsValue document, PatcherOperationScope scope)
+        {
+            if (collectionName == null)
+                ThrowLoadParameterIsMandatory(nameof(collectionName));
+            if (document == null)
+                ThrowLoadParameterIsMandatory(nameof(document));
+
+            var transformed = scope.ToBlittable(document.AsObject());
+
+            string prefixedId = null;
+
+            if (_script.IsLoadingToDefaultCollection == false ||
+                _script.NonDefaultCollections.Length > 0 && 
+                _script.DefaultCollection.Equals(collectionName, StringComparison.OrdinalIgnoreCase) == false)
+            {
+                DynamicJsonValue metadata;
+
+                if (transformed[Constants.Documents.Metadata.Key] != null)
+                    metadata = transformed[Constants.Documents.Metadata.Key] as DynamicJsonValue;
+                else
+                    transformed[Constants.Documents.Metadata.Key] = metadata = new DynamicJsonValue();
+
+                prefixedId = GetPrefixedId(_currentlyTransformed.DocumentKey, collectionName);
+
+                metadata[Constants.Documents.Metadata.Collection] = collectionName;
+                metadata[Constants.Documents.Metadata.Id] = prefixedId;
+            }
+
+            var id = prefixedId ?? _currentlyTransformed.DocumentKey;
+
+            var transformResult = Context.ReadObject(transformed, id);
+
+            _commands.Add(new PutCommandDataWithBlittableJson(id, null, transformResult));
+        }
+
+        private string GetPrefixedId(string documentId, string loadCollectionName)
+        {
+            return $"{documentId}/{_script.IdPrefixForCollection[loadCollectionName]}/";
```

Comment:
@ayende Considering my latest comment here, do you still think we should provide users with an option to define their own ids in ETL scripts? 

Note we are already able to load such docs using `LoadStartingWith`:

https://github.com/ravendb/ravendb/blob/v4.0/test/SlowTests/Server/Documents/ETL/Raven/BasicRavenEtlTests.cs#L436

https://github.com/ravendb/ravendb/blob/v4.0/test/SlowTests/Server/Documents/ETL/Raven/BasicRavenEtlTests.cs#L282